Various *.rego inputs that cause panics

All of these are reproducible when running `--serve` instead of REPL.

### File with single `.`

```
$ opa run
OPA 0.9.1 (commit d5ba22b7, built at 2018-08-16T15:46:52Z)

Run 'help' to see a list of commands.

> .
| 
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/open-policy-agent/opa/ast.newParserErrorDetail(0xc4201d3050, 0x2, 0x8, 0x2, 0x0, 0x1, 0x0)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:708 +0x16f
github.com/open-policy-agent/opa/ast.formatParserError(0x0, 0x0, 0xc4201d3050, 0x2, 0x8, 0xc4201c2d20, 0xc4202976b0)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:521 +0x19f
github.com/open-policy-agent/opa/ast.formatParserErrors(0x0, 0x0, 0xc4201d3050, 0x2, 0x8, 0x1709500, 0xc4201d5940, 0x2, 0x0)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:505 +0x135
github.com/open-policy-agent/opa/ast.ParseStatements(0x0, 0x0, 0xc4201d3048, 0x2, 0x1, 0xc4201d3048, 0x2, 0xc420297900, 0x1040845, 0xc4201d5420, ...)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:470 +0x6af
github.com/open-policy-agent/opa/repl.(*REPL).evalBufferMulti(0xc4201d0900, 0x170e9c0, 0xc4200220c0, 0x1, 0x2)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:663 +0x12e
github.com/open-policy-agent/opa/repl.(*REPL).OneShot(0xc4201d0900, 0x170e9c0, 0xc4200220c0, 0xc4201d3044, 0x0, 0x0, 0x0)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:278 +0x1be
github.com/open-policy-agent/opa/repl.(*REPL).Loop(0xc4201d0900, 0x170e9c0, 0xc4200220c0)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:178 +0x303
github.com/open-policy-agent/opa/runtime.(*Runtime).StartREPL(0xc4201a4ea0, 0x170e9c0, 0xc4200220c0)
	/go/src/github.com/open-policy-agent/opa/runtime/runtime.go:289 +0x1d7
github.com/open-policy-agent/opa/cmd.init.4.func1(0xc42016e480, 0x1a4af00, 0x0, 0x0)
	/go/src/github.com/open-policy-agent/opa/cmd/run.go:122 +0x4fd
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).execute(0xc42016e480, 0x1a4af00, 0x0, 0x0, 0xc42016e480, 0x1a4af00)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:651 +0x23d
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1a24040, 0xc4201d9f68, 0x1533055, 0xc4201d9f78)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:726 +0x2d4
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).Execute(0x1a24040, 0x2, 0xc420161da0)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:685 +0x2b
main.main()
	/go/src/github.com/open-policy-agent/opa/main.go:12 +0x2d
```

### File with single `-`

```
$ opa run
OPA 0.9.1 (commit d5ba22b7, built at 2018-08-16T15:46:52Z)

Run 'help' to see a list of commands.

> -
| 
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/open-policy-agent/opa/ast.newParserErrorDetail(0xc4201dd068, 0x2, 0x8, 0x2, 0x0, 0x1, 0x0)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:708 +0x16f
github.com/open-policy-agent/opa/ast.formatParserError(0x0, 0x0, 0xc4201dd068, 0x2, 0x8, 0xc4201cccd0, 0xc4202a36b0)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:521 +0x19f
github.com/open-policy-agent/opa/ast.formatParserErrors(0x0, 0x0, 0xc4201dd068, 0x2, 0x8, 0x1709500, 0xc4201dee00, 0x2, 0x0)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:505 +0x135
github.com/open-policy-agent/opa/ast.ParseStatements(0x0, 0x0, 0xc4201dd060, 0x2, 0x1, 0xc4201dd060, 0x2, 0xc4202a3900, 0x1040845, 0xc4201de900, ...)
	/go/src/github.com/open-policy-agent/opa/ast/parser_ext.go:470 +0x6af
github.com/open-policy-agent/opa/repl.(*REPL).evalBufferMulti(0xc4201da7e0, 0x170e9c0, 0xc4200220c0, 0x1, 0x2)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:663 +0x12e
github.com/open-policy-agent/opa/repl.(*REPL).OneShot(0xc4201da7e0, 0x170e9c0, 0xc4200220c0, 0xc4201dd04c, 0x0, 0x0, 0x0)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:278 +0x1be
github.com/open-policy-agent/opa/repl.(*REPL).Loop(0xc4201da7e0, 0x170e9c0, 0xc4200220c0)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:178 +0x303
github.com/open-policy-agent/opa/runtime.(*Runtime).StartREPL(0xc4201aed80, 0x170e9c0, 0xc4200220c0)
	/go/src/github.com/open-policy-agent/opa/runtime/runtime.go:289 +0x1d7
github.com/open-policy-agent/opa/cmd.init.4.func1(0xc420206240, 0x1a4af00, 0x0, 0x0)
	/go/src/github.com/open-policy-agent/opa/cmd/run.go:122 +0x4fd
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).execute(0xc420206240, 0x1a4af00, 0x0, 0x0, 0xc420206240, 0x1a4af00)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:651 +0x23d
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1a24040, 0xc4201e3f68, 0x1533055, 0xc4201e3f78)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:726 +0x2d4
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).Execute(0x1a24040, 0x2, 0xc42018fd40)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:685 +0x2b
main.main()
	/go/src/github.com/open-policy-agent/opa/main.go:12 +0x2d
```

I think it would generally be a good idea to run something like `go-fuzz` against this.

If I find more, I will report it.

--- a/ast/parser_ext.go
+++ b/ast/parser_ext.go
@@ -702,18 +702,56 @@ type parserErrorDetail struct {
 }
 
 func newParserErrorDetail(bs []byte, pos position) *parserErrorDetail {
-	lines := strings.Split(string(bs), "\n")
-	line := lines[pos.line-1]
-	idx := pos.col - 1
-	for (idx >= len(line) || unicode.IsSpace(rune(line[idx]))) && idx > 0 {
-		idx--
+
+	offset := pos.offset
+
+	// Find first non-space character at or before offset position.
+	if offset >= len(bs) {
+		offset = len(bs) - 1
+	} else if offset < 0 {
+		offset = 0
 	}
+
+	for offset > 0 && unicode.IsSpace(rune(bs[offset])) {
+		offset--
+	}
+
+	// Find beginning of line containing offset.
+	begin := offset
+
+	for begin > 0 && !isNewLineChar(bs[begin]) {
+		begin--
+	}
+
+	if isNewLineChar(bs[begin]) {
+		begin++
+	}
+
+	// Find end of line containing offset.
+	end := offset
+
+	for end < len(bs) && !isNewLineChar(bs[end]) {
+		end++
+	}
+
+	if begin > end {
+		begin = end
+	}
+
+	// Extract line and compute index of offset byte in line.
+	line := bs[begin:end]
+	index := offset - begin
+
 	return &parserErrorDetail{
-		line: lines[pos.line-1],
-		idx:  idx,
+		line: string(line),
+		idx:  index,
 	}
 }
 
 func (d parserErrorDetail) Lines() []string {
 	return []string{d.line, strings.Repeat(" ", d.idx) + "^"}
+}
+
+func isNewLineChar(b byte) bool {
+	return b == '\r' || b == '\n'
 }

--- a/ast/parser_test.go
+++ b/ast/parser_test.go
@@ -1420,6 +1420,16 @@ func TestParseErrorDetails(t *testing.T) {
 		input string
 	}{
 		{
+			note: "no match: bad rule name",
+			exp: &parserErrorDetail{
+				line: ".",
+				idx:  0,
+			},
+			input: `
+package test
+.`,
+		},
+		{
 			note: "no match: bad termination for comprehension",
 			exp: &parserErrorDetail{
 				line: "p = [true | true}",

